src/go/rpk/pkg/cli/topic/trim.go

```diff
@@ -0,0 +1,279 @@
+				o, err = parseOffsetFile(fs, fromFile)
+				out.MaybeDie(err, "unable to parse file %q: %v", fromFile, err)
+			} else {
+				if len(args) == 0 {
+					out.Die("Error: required 'topic' not set; Usage: 'rpk topic trim-prefix [TOPIC]'")
+				}
+				if !all && offset == "" {
+					out.Die("Error: provide one of ['--offset', '--to-hwm']")
+				}
+				topic := args[0]
+				o, err = parseOffsetArgs(cmd.Context(), adm, all, topic, offset, partitions)
+				out.MaybeDieErr(err)
+			}
+			if !noConfirm {
+				err := printDeleteRecordRequest(cmd.Context(), adm, o)
+				out.MaybeDie(err, "unable to print trimming request: %v", err)
+				fmt.Println()
+				confirmed, err := out.Confirm("Confirm trimming of the above offsets?")
+				out.MaybeDie(err, "unable to confirm: %v", err)
+				if !confirmed {
+					out.Exit("Trimming canceled.")
+				}
+				fmt.Println()
+			}
+			drr, err := adm.DeleteRecords(cmd.Context(), o)
+			out.MaybeDie(err, "unable to trim records: %v", err)
+			printDeleteRecordResponse(drr)
+		},
+	}
+
+	cmd.Flags().StringVarP(&offset, "offset", "o", "", "Offset to set the partition's low watermark (start offset) to")
+	cmd.Flags().IntSliceVarP(&partitions, "partitions", "p", nil, "Comma-separated list of partitions to trim records from (default to all)")
+	cmd.Flags().BoolVar(&all, "to-hwm", false, "Set the partition's low watermark to the partition's high watermark (trim all)")
+	cmd.Flags().StringVarP(&fromFile, "from-file", "f", "", "File of topic/partition/offset for which to trim offsets for")
+	cmd.Flags().BoolVar(&noConfirm, "no-confirm", false, "Disable confirmation prompt")
+
+	cmd.MarkFlagsMutuallyExclusive("to-hwm", "offset")
+	cmd.MarkFlagsMutuallyExclusive("from-file", "offset")
+	cmd.MarkFlagsMutuallyExclusive("from-file", "partitions")
+
+	return cmd
+}
+
+// parseOffsetArgs creates the kadm.Offsets based on the passed args via rpk
+// flags. It checks if the topic exists and defaults to all partitions in the
+// case that partitions were not provided.
+func parseOffsetArgs(ctx context.Context, adm *kadm.Client, all bool, topic, offset string, partitions []int) (kadm.Offsets, error) {
+	// First we check if the topic exists
+	listed, err := adm.ListTopics(ctx, topic)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list topic %q metadata: %v", topic, err)
+	}
+	listed.EachError(func(d kadm.TopicDetail) {
+		err = fmt.Errorf("unable to list the topic %q: %v", topic, d.Err)
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// If no partitions are provided, we trim the record to the
+	// provided offset in all partitions.
+	if len(partitions) == 0 {
+		listed.EachPartition(func(pd kadm.PartitionDetail) {
+			partitions = append(partitions, int(pd.Partition))
+		})
+	}
+
+	o := make(kadm.Offsets)
+	for _, p := range partitions {
+		var parsedOffset int64
+		if all {
+			parsedOffset = -1 // -1 maps to the partition's high watermark.
+		} else {
+			parsedOffset, err = parseOffsetFlag(ctx, adm, topic, offset, p)
```

Comment:
This function issues a metadata request for _every_ individual partition, and calculates time for every partition. This should instead take an input set of topics/partitions and only do the calculation & metadata request once.

Reply:
Assuming by `This function` refers to `parseOffsetFlag` function:

This is true only for @-prefixed offsets (for numeric offsets, it just converts from string to int) and when all partitions are selected (or no --partition flag was specified, which is a design choice); We issue a `kadm.ListOffsetsAfterMilli` which I think it's what you refer to metadata + calculation.